File: frontend/lib/api/scanner-status.ts
import { apiFetch } from "./client";

export type BinaryChecklistItem = {
  os: string;
  arch: string;
  published: boolean;
  version: string | null;
  uploaded_at: string | null;
  uploaded_by: string | null;
};

export type RunningScanItem = {
  scan_id: string;
  project_id: string;
  started_at: string | null;
  stuck: boolean;
};

export type QueueStatus = {
  running: number;
  queued: number;
  max_concurrent: number;
  running_scans: RunningScanItem[];
};

export type FailureItem = {
  scan_id: string;
  project_id: string;
  scan_type: string;
  error_message: string | null;
  completed_at: string | null;
};

export type ScannerStatus = {
  engine_available: boolean;
  binaries: BinaryChecklistItem[];
  queue: QueueStatus;
  recent_failures: FailureItem[];
};

export function getScannerStatus() {
  return apiFetch<ScannerStatus>("/admin/scanner-status");
}

File: backend/app/schemas/scanner_status.py
from datetime import datetime

from pydantic import BaseModel


class BinaryChecklistItem(BaseModel):
    os: str
    arch: str
    published: bool
    version: str | None = None
    uploaded_at: datetime | None = None
    uploaded_by: str | None = None


class RunningScanItem(BaseModel):
    scan_id: str
    project_id: str
    started_at: datetime | None
    stuck: bool


class QueueStatus(BaseModel):
    running: int
    queued: int
    max_concurrent: int
    running_scans: list[RunningScanItem]


class FailureItem(BaseModel):
    scan_id: str
    project_id: str
    scan_type: str
    error_message: str | None
    completed_at: datetime | None


class ScannerStatusResponse(BaseModel):
    engine_available: bool
    binaries: list[BinaryChecklistItem]
    queue: QueueStatus
    recent_failures: list[FailureItem]
